lib/teleterm/services/desktop/directorysharing.go

```diff
@@ -0,0 +1,313 @@
+func (d *DirectoryAccess) Stat(relativePath string) (*FileOrDirInfo, error) {
+	path, err := d.getSafePath(relativePath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	stat, err := os.Stat(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	info, err := d.readFileOrDirInfo(relativePath, stat)
+	return info, trace.Wrap(err)
+}
+
+// ReadDir lists files and directories within the given directory path, skips symlinks.
+func (d *DirectoryAccess) ReadDir(relativePath string) ([]*FileOrDirInfo, error) {
+	path, err := d.getSafePath(relativePath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	entries, err := os.ReadDir(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var results []*FileOrDirInfo
+	for _, entry := range entries {
+		fileInfo, err := entry.Info()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// Skip symlinks, we can't present them properly in the remote machine.
+		if fileInfo.Mode().Type()&os.ModeSymlink != 0 {
+			continue
+		}
+
+		entryRelativePath := filepath.Join(relativePath, fileInfo.Name())
+		fileOrDir, err := d.readFileOrDirInfo(entryRelativePath, fileInfo)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		results = append(results, fileOrDir)
+	}
+
+	return results, nil
+}
+
+// Read reads a slice of a file into buf. Returns the number of read bytes.
+func (d *DirectoryAccess) Read(relativePath string, offset int64, buf []byte) (n int, err error) {
+	path, err := d.getSafePath(relativePath)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	defer func() {
+		closeErr := file.Close()
+		if err == nil {
+			// Only update err if no previous error occurred.
+			err = trace.Wrap(closeErr)
+		}
+	}()
+
+	_, err = file.Seek(offset, io.SeekStart)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	n, err = file.Read(buf)
+	if err != nil && !errors.Is(err, io.EOF) {
+		return 0, trace.Wrap(err)
+	}
+	return n, trace.Wrap(err)
+}
+
+// Write writes data to a file at a given offset.
+func (d *DirectoryAccess) Write(relativePath string, offset int64, data []byte) (n int, err error) {
+	path, err := d.getSafePath(relativePath)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	defer func() {
+		closeErr := file.Close()
+		if err == nil {
+			// Only update err if no previous error occurred.
+			err = trace.Wrap(closeErr)
+		}
+	}()
+
+	_, err = file.Seek(offset, io.SeekStart)
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	n, err = file.Write(data)
+	return n, trace.Wrap(err)
+}
+
+// Truncate truncates a file to the specified size.
+func (d *DirectoryAccess) Truncate(relativePath string, size int64) error {
+	path, err := d.getSafePath(relativePath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.Wrap(os.Truncate(path, size))
```

Comment:
```suggestion
	return trace.ConvertSystemError(os.Truncate(path, size))
```

For filesystem operations I think `ConvertSystemError` will try to find the most appropriate trace error variant.

You might consider doing the same for other write/close/read calls.

Up to you if you want go to through with this - it doesn't _really_ matter to us what type of error it ends up being since we treat them all the same, but maybe it might result in the user seeing a slightly better error message.

Reply:
Thanks, I added `ConvertSystemError` for the file system operations.